pysmt/smtlib/parser/parser.py

```diff
@@ -1172,14 +1171,12 @@ def _cmd_define_fun(self, current, tokens):
         rtype = self.parse_type(tokens, current)
 
         for (x,t) in namedparams:
-            v = self._get_var(x, t)
+            v = self.env.formula_manager.FreshSymbol(typename=t,
+                                                        template="__"+x+"%d")
             self.cache.bind(x, v)
             formal.append(v)
         # Parse expression using also parameters
         ebody = self.get_expression(tokens)
-        # Discard parameters
```

Comment:
Why did you remove this piece of code? I think it is still needed to avoid catching an undefined variable outside the define-fun...

Reply:
What do you mean?

For example:
```
(define-fun f ((n Int)) Int n)
(assert (= n n))
```

Isn't the `n` in the second line undefined? Shouldn't this be catched?
Or did you mean another scenario?

Reply:
Exactly. I think (but I haven't tested this) that if you remove this code, the parsing of your example would succeed, because n is bound to a symbol by the declare-fun (you are asserting "Equals(__n0, __n0)").

I just noticed that the code needs to be modified a bit to remove the correct binding name (because currently it relays on the fact that the symbol name is the same as the binding name, but here we are changing this):
```python
bindings = []
for (x,t) in namedparams:
￼    v = self.env.formula_manager.FreshSymbol(typename=t, template="__"+x+"%d")
￼    self.cache.bind(x, v)
￼    formal.append(v) 
    bindings.append(x) # Remember the name
# Parse expression using also parameters
￼ebody = self.get_expression(tokens)
￼# Discard parameters
for x in bindings:
    self.cache.unbind(x) 
```
This should work.